pdns/channel.hh

```diff
@@ -0,0 +1,329 @@
+     */
+    int getDescriptor() const;
+    /**
+     * \brief Whether the remote end has closed the channel.
+     */
+    bool isClosed() const
+    {
+      return d_closed;
+    }
+
+  private:
+    FDWrapper d_fd;
+    bool d_closed{false};
+    bool d_throwOnEOF{true};
+  };
+
+  /**
+   * \brief Create a channel to notify one thread from another one, accepting multiple senders and receivers.
+   *
+   * \return A pair of Notifier and Sender objects.
+   *
+   * \throw runtime_error if the channel creation failed.
+   */
+  std::pair<Notifier, Waiter> createNotificationQueue(bool nonBlocking = true, size_t pipeBufferSize = 0, bool throwOnEOF = true);
+
+  template <typename T, typename D>
+  bool Sender<T, D>::send(std::unique_ptr<T, D>&& object) const
+  {
+    // we do not release right away because we might need the custom deleter later
+    auto ptr = object.get();
+    static_assert(sizeof(ptr) <= PIPE_BUF, "Writes up to PIPE_BUF are guaranted not to interleaved and to either fully succeed or fail");
+    while (true) {
+#if __SANITIZE_THREAD__
+      __tsan_release(ptr);
+#endif /* __SANITIZE_THREAD__ */
+      ssize_t sent = write(d_fd.getHandle(), &ptr, sizeof(ptr));
+
+      if (sent == sizeof(ptr)) {
+        // we cannot touch it anymore
+        object.release();
+        return true;
+      }
+      else {
+#if __SANITIZE_THREAD__
+        __tsan_acquire(ptr);
+#endif /* __SANITIZE_THREAD__ */
+        if (errno == EINTR) {
+          continue;
+        }
+        if (errno == EAGAIN || errno == EWOULDBLOCK) {
+          return false;
+        }
+        else {
+          throw std::runtime_error("Unable to write to channel:" + stringerror());
+        }
+      }
+    }
+  }
+
+  template <typename T, typename D>
+  void Sender<T, D>::close()
+  {
+    d_fd.reset();
+  }
+
+  template <typename T, typename D>
+  std::optional<std::unique_ptr<T, D>> Receiver<T, D>::receive()
+  {
+    return receive(std::default_delete<T>());
+  }
+
+  template <typename T, typename D>
+  std::optional<std::unique_ptr<T, D>> Receiver<T, D>::receive(D deleter)
+  {
+    while (true) {
+      std::optional<std::unique_ptr<T, D>> result;
+      T* obj{nullptr};
+      ssize_t got = read(d_fd.getHandle(), &obj, sizeof(obj));
+      if (got == sizeof(obj)) {
+#if __SANITIZE_THREAD__
+        __tsan_acquire(obj);
+#endif /* __SANITIZE_THREAD__ */
+        return std::unique_ptr<T, D>(obj, deleter);
+      }
+      else if (got == 0) {
+        d_closed = true;
+        if (!d_throwOnEOF) {
+          return result;
+        }
+        throw std::runtime_error("EOF while reading from Channel receiver");
+      }
+      else if (got == -1) {
+        if (errno == EINTR) {
+          continue;
+        }
+        if (errno == EAGAIN || errno == EWOULDBLOCK) {
+          return result;
+        }
+        throw std::runtime_error("Error while reading from Channel receiver: " + stringerror());
+      }
+      else {
```

Comment:
Does it make sense to test for write returning != writesize on the write side as well? It *should* never happen, but still...

Reply:
I think we do, don't we?

Reply:
Ah right, just the zero case is not handled as a separate case. errno is not set if write returns zero, so to be pedantic the zero case should be handled separately,.

Reply:
Agreed, I'll do that!